userspace/libsinsp/test/async_key_value_source.ut.cpp

```diff
@@ -58,47 +99,103 @@ TEST(async_key_value_source_test, no_lookup)
 TEST(async_key_value_source_test, basic)
 {
 	std::unique_ptr<test_key_value_source> t(new test_key_value_source(0, UINT64_MAX));
-	uint64_t value;
+	result res;
 
-	while(!t->lookup("1", value))
+	while(!t->lookup("1", res))
 	{
-		sleep(1);
+		std::this_thread::sleep_for(std::chrono::milliseconds(100));
 	}
-	ASSERT_EQ(1, value);
+	ASSERT_EQ(1, res.val);
 }
 
 TEST(async_key_value_source_test, long_delay_lookups)
 {
-	std::unique_ptr<test_key_value_source> t(new test_key_value_source(5, UINT64_MAX));
-	uint64_t value;
+	std::unique_ptr<test_key_value_source> t(new test_key_value_source(500, UINT64_MAX));
+	result res;
 
-	while(!t->lookup("1", value))
+	while(!t->lookup("1", res))
 	{
-		sleep(1);
+		std::this_thread::sleep_for(std::chrono::milliseconds(100));
 	}
-	ASSERT_EQ(1, value);
+	ASSERT_EQ(1, res.val);
 }
 
 TEST(async_key_value_source_test, basic_nowait)
 {
 	std::unique_ptr<test_key_value_source> t(new test_key_value_source(0, 0));
-	uint64_t value;
+	result res;
 
-	while(!t->lookup("1", value))
+	while(!t->lookup("1", res))
 	{
-		sleep(1);
+		std::this_thread::sleep_for(std::chrono::milliseconds(100));
 	}
-	ASSERT_EQ(1, value);
+	ASSERT_EQ(1, res.val);
 }
 
 TEST(async_key_value_source_test, long_delay_lookups_nowait)
 {
-	std::unique_ptr<test_key_value_source> t(new test_key_value_source(5, 0));
-	uint64_t value;
+	std::unique_ptr<test_key_value_source> t(new test_key_value_source(500, 0));
+	result res;
 
-	while(!t->lookup("1", value))
+	while(!t->lookup("1", res))
 	{
-		sleep(1);
+		std::this_thread::sleep_for(std::chrono::milliseconds(100));
 	}
-	ASSERT_EQ(1, value);
+	ASSERT_EQ(1, res.val);
+}
+
+TEST(async_key_value_source_test, async)
+{
+	uint64_t ttl_ms = std::numeric_limits<uint64_t>::max();
+	short num_failures = 3;
+	test_key_value_source t(0, 0, ttl_ms, num_failures);
+	result res;
+	std::condition_variable cv;
+	std::mutex cv_m;
```

Comment:
maybe a [std::future](https://en.cppreference.com/w/cpp/thread/future) would be a better fit here

Reply:
yep, probably I could have used [std::promise](https://en.cppreference.com/w/cpp/thread/promise), but I think this is fine too.